protocol/consistencychecks.go

```diff
@@ -0,0 +1,379 @@
+	if !ap.Verify(str.TreeHash) {
+		return invalidProof, ErrorBadAuthPath
+	}
+
+	return proofType, nil
+}
+
+func (cc *ConsistencyChecks) verifySTR(str *m.SignedTreeRoot) error {
+	// verify STR's signature
+	if !cc.signKey.Verify(str.Serialize(), str.Signature) {
+		return ErrorBadSignature
+	}
+
+	// verify hash chain
+	// TODO: clearly verify which is the client's actual expectation.
+	// See #81
+	if (str.Epoch == cc.SavedSTR.Epoch && bytes.Equal(cc.SavedSTR.Signature, str.Signature)) ||
+		(str.Epoch == cc.SavedSTR.Epoch+1 && str.VerifyHashChain(cc.SavedSTR.Signature)) {
+		return nil
+	}
+
+	// TODO: verify the directory's policies as well. See #115
+	return ErrorBadSTR
+}
+
+// verifyFulfilledPromise verifies issued TBs was inserted
+// in the directory as promised.
+func (cc *ConsistencyChecks) verifyFulfilledPromise(uname string,
+	str *m.SignedTreeRoot,
+	ap *m.AuthenticationPath) error {
+	if !cc.useTBs {
+		return nil
+	}
+
+	for name, tb := range cc.TBs {
+		if cc.SavedSTR.Epoch+1 == str.Epoch &&
+			name == uname {
+			if !tb.Verify(ap) {
+				return ErrorBrokenPromise
+			}
+			delete(cc.TBs, uname)
+		} else if cc.SavedSTR.Epoch < str.Epoch {
+			// TODO: should verify when we introduce KeyLookupInEpoch
+			// and monitoring with missed epochs,
+			// see: https://github.com/coniks-sys/coniks-go/pull/74#discussion_r84930999
+			// and: https://github.com/coniks-sys/coniks-go/pull/74#discussion_r84501211
+			delete(cc.TBs, name)
+		}
+		// keep current epoch's returned promises
+		// for future verifications
+	}
+	return nil
+}
+
+// verifyReturnedPromise validates the returned promise
+// based on the request type. Note that the directory
+// returns a promise iff the returned proof is
+// _a proof of absence_.
+// 	If the request is a registration, and
+// 	- the request is successful, then the directory should return a promise for the new binding.
+// 	- the request is failed because of ErrorNameExisted, then the directory should return a promise for that existed binding.
+//
+// 	If the request is a key lookup, and
+// 	- the request is successful, then the directory should return a promise for the lookup binding.
+func (cc *ConsistencyChecks) verifyReturnedPromise(requestType int,
+	responseCode ErrorCode, df *DirectoryProof,
+	uname string, key []byte) error {
+	if !cc.useTBs {
+		return nil
+	}
+
+	ap := df.AP
+	str := df.STR
+	tb := df.TB
+
+	// the client should receive a signed promise iff
+	// the directory returns a proof of absence
+	if bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
+		if tb != nil {
+			// the directory returned a proof of inclusion
+			// _and_ a promise
+			return ErrorMalformedDirectoryMessage
+		}
+		return nil
+	}
+
+	switch requestType {
+	case RegistrationType:
+		if responseCode == ErrorNameExisted {
+			// ignore the key value verification since we may have no information about that existed binding.
+			if verifResult := cc.verifyTB(tb, uname, nil, str, ap); verifResult != nil {
```

Comment:
Can we rename `verifResult` to something that indicates that this means that the verification failed (like `verifError` or just `err)? I think this would slightly increase the readability (here and everywhere else this var name is used for this particular purpose).